pkg/dependency/parser/nodejs/pnpm/parse.go

```diff
@@ -107,6 +141,97 @@ func (p *Parser) parse(lockVer float64, lockFile LockFile) ([]ftypes.Package, []
 	return pkgs, deps
 }
 
+func (p *Parser) parseV9(lockFile LockFile) ([]ftypes.Package, []ftypes.Dependency) {
+	lockVer := 9.0
+	resolvedPkgs := make(map[string]ftypes.Package)
+	resolvedDeps := make(map[string]ftypes.Dependency)
+
+	// Check all snapshots and save with resolved versions
+	resolvedSnapshots := make(map[string][]string)
+	for depPath, snapshot := range lockFile.Snapshots {
+		name, version := p.parseDepPath(depPath, lockVer)
+
+		var dependsOn []string
+		for depName, depVer := range lo.Assign(snapshot.OptionalDependencies, snapshot.Dependencies) {
+			depVer = p.trimPeerDeps(depVer, lockVer) // pnpm has already separated dep name. therefore, we only need to separate peer deps.
+			depVer = p.parseVersion(depPath, depVer, lockVer)
+			id := packageID(depName, depVer)
+			if _, ok := lockFile.Packages[id]; ok {
+				dependsOn = append(dependsOn, id)
+			}
+		}
+		if dependsOn != nil {
+			resolvedSnapshots[packageID(name, version)] = dependsOn
+		}
+
+	}
+
+	for depPath, pkgInfo := range lockFile.Packages {
+		name, ver := p.parseDepPath(depPath, lockVer)
+		parsedVer := p.parseVersion(depPath, ver, lockVer)
+
+		if pkgInfo.Version != "" {
+			parsedVer = pkgInfo.Version
+		}
+
+		// By default, pkg is dev pkg.
+		// We will update `Dev` field later.
+		dev := true
+		relationship := ftypes.RelationshipIndirect
+		if dep, ok := lockFile.Importers.Root.DevDependencies[name]; ok && dep.Version == ver {
+			relationship = ftypes.RelationshipDirect
+		}
+		if dep, ok := lockFile.Importers.Root.Dependencies[name]; ok && dep.Version == ver {
+			relationship = ftypes.RelationshipDirect
+			dev = false // mark root direct deps to update `dev` field of their child deps.
+		}
+
+		id := packageID(name, parsedVer)
+		resolvedPkgs[id] = ftypes.Package{
+			ID:           id,
+			Name:         name,
+			Version:      parsedVer,
+			Relationship: relationship,
+			Dev:          dev,
+		}
+
+		// Save child deps
+		if dependsOn, ok := resolvedSnapshots[depPath]; ok {
+			sort.Strings(dependsOn)
+			resolvedDeps[id] = ftypes.Dependency{
+				ID:        id,
+				DependsOn: dependsOn, // Deps from dependsOn has been resolved when parsing snapshots
+			}
+		}
+	}
+
+	// Overwrite the `Dev` field for dev deps and their child dependencies.
+	for _, pkg := range resolvedPkgs {
+		if !pkg.Dev {
+			p.markRootPkgs(pkg.ID, resolvedPkgs, resolvedDeps)
+		}
+	}
+
+	return maps.Values(resolvedPkgs), maps.Values(resolvedDeps)
+}
+
+// markRootPkgs sets `Dev` to false for non dev dependency.
+func (p *Parser) markRootPkgs(id string, libs map[string]ftypes.Package, deps map[string]ftypes.Dependency) {
+	lib, ok := libs[id]
```

Comment:
```suggestion
	pkg, ok := pkgs[id]
```

Reply:
fixed in https://github.com/aquasecurity/trivy/pull/6617/commits/1ebe152d278656cbab059a9678d7d359d65ddd9c